pgxn/neon/walsender_hooks.c

```diff
@@ -0,0 +1,121 @@
+#include "walsender_hooks.h"
+#include "postgres.h"
+#include "fmgr.h"
+#include "access/xlogdefs.h"
+#include "replication/walsender.h"
+#include "access/xlog.h"
+#include "access/xlog_internal.h"
+#include "access/xlogreader.h"
+#include "miscadmin.h"
+#include "utils/wait_event.h"
+
+#include "neon_walreader.h"
+#include "walproposer.h"
+
+static NeonWALReader *wal_reader = NULL;
+extern XLogRecPtr WalSndWaitForWal(XLogRecPtr loc);
+extern bool GetDonorShmem(XLogRecPtr *donor_lsn);
+
+static XLogRecPtr NeonWALReadWaitForWAL(XLogRecPtr loc)
+{
+    while(!NeonWALReaderUpdateDonor(wal_reader))
+        pg_usleep(1000);
+
+    return WalSndWaitForWal(loc);
+}
+
+static int NeonWALPageRead(
+    XLogReaderState *xlogreader,
+    XLogRecPtr targetPagePtr,
+    int reqLen,
+    XLogRecPtr targetRecPtr,
+    char *readBuf)
+{
+    // Wait for flush pointer to advance past our request
+    XLogRecPtr flushptr = NeonWALReadWaitForWAL(targetPagePtr + reqLen);
+    int count;
+    if (flushptr < targetPagePtr + reqLen)
+	return -1;
+
+    // Read at most XLOG_BLCKSZ bytes
+    if (targetPagePtr + XLOG_BLCKSZ <= flushptr)
+	count = XLOG_BLCKSZ;
+    else
+	count = flushptr - targetPagePtr;
+
+    // Sometimes walsender requests non-monotonic sequences of WAL. If that's
+    // the case, we have to reset streaming from remote at the correct position.
+    if(targetPagePtr != NeonWALReaderGetRemLsn(wal_reader))
+    {
+	NeonWALReaderResetRemote(wal_reader);
+    }
+
+    static WaitEventSet *wait_event_set;
+    for(;;)
+    {
+	NeonWALReadResult res = NeonWALRead(
+	    wal_reader,
+	    readBuf,
+	    targetPagePtr,
+	    count,
+	    walprop_pg_get_timeline_id());
+	if(res == NEON_WALREAD_SUCCESS)
+	{
+            // Free event set, since we'll have to recreate it for a new socket later anyway
+	    if (NeonWALReaderEvents(wal_reader) == 0 && wait_event_set != NULL)
+	    {
+		FreeWaitEventSet(wait_event_set);
+		wait_event_set = NULL;
+	    }
+            // Required by xlogreader
```

Comment:
Similarly to neon_segment_open/close hooks seems like this is needed

Reply:
Yes, but I think it makes sense to respect the invariants outlined in the comment in XLogReaderRoutine

Reply:
Not sure. Again, in case of remote read we don't really open the segment. If you leave it I'd suggest to at least add a comment that we don't use these fields.

Reply:
Ok, added a comment.